src/violation.rs

```diff
@@ -77,22 +20,28 @@ impl ViolationImpl {
             violations: vec![],
         }
     }
-}
 
-impl Violations for ViolationImpl {
-    fn add_violation(&mut self, violation: ViolationImpl) {
-        self.violations.push(violation);
-    }
-
-    fn is_valid(&self) -> bool {
-        self.violations.is_empty()
+    pub fn with_violations<A: AsRef<str>>(
+        constraint: A,
+        code: A,
+        message: A,
+        violations: Vec<Violation>,
+    ) -> Self {
+        Self {
+            constraint: constraint.as_ref().to_owned(),
+            code: code.as_ref().to_owned(),
+            message: message.as_ref().to_owned(),
+            violations,
+        }
     }
 
-    fn violations(&self) -> &[ViolationImpl] {
+    pub fn violations(&self) -> &[Violation] {
         &self.violations
     }
+}
 
-    fn checkpoint(&self) -> Checkpoint {
-        Checkpoint::new(self.violations.len())
+impl fmt::Display for Violation {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "A validation error occurred: {}", self.message)
```

Comment:
`fmt::Display` is analogous to Java's `toString`; I think it should produce a minimal text representation of the `Violation` rather than a sentence. Its caller would provide the sentence.

What's our use case for this impl at the moment? Could we get away with just deriving `Debug` and using `{:?}` in our format strings?

Reply:
Added `Display` because it was was required  to be implemented by `Error` bound.

Reply:
Ah, that makes sense. Thanks for explaining. We can punt on changing the format.